libs/vec/src/main21/java/org/elasticsearch/vec/internal/IndexInputUtils.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.vec.internal;
+
+import org.apache.lucene.store.FilterIndexInput;
+import org.apache.lucene.store.IndexInput;
+
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+
+public final class IndexInputUtils {
+
+    static final Class<?> MSINDEX_CLS, MS_MSINDEX_CLS;
+    static final VarHandle SEGMENTS_ARRAY, CHUNK_SIZE_POWER, CHUNK_SIZE_MASK, MULTI_OFFSET;
+
+    static {
+        try {
+            MSINDEX_CLS = Class.forName("org.apache.lucene.store.MemorySegmentIndexInput");
+            MS_MSINDEX_CLS = Class.forName("org.apache.lucene.store.MemorySegmentIndexInput$MultiSegmentImpl");
+            var lookup = privilegedPrivateLookupIn(MSINDEX_CLS, MethodHandles.lookup());
+            SEGMENTS_ARRAY = privilegedFindVarHandle(lookup, MSINDEX_CLS, "segments", MemorySegment[].class);
+            CHUNK_SIZE_POWER = privilegedFindVarHandle(lookup, MSINDEX_CLS, "chunkSizePower", int.class);
+            CHUNK_SIZE_MASK = privilegedFindVarHandle(lookup, MSINDEX_CLS, "chunkSizeMask", long.class);
+            MULTI_OFFSET = privilegedFindVarHandle(lookup, MS_MSINDEX_CLS, "offset", long.class);
+        } catch (ClassNotFoundException e) {
+            throw new AssertionError(e);
+        } catch (IllegalAccessException e) {
+            throw new AssertionError("should not happen, check opens", e);
+        } catch (PrivilegedActionException e) {
+            throw new AssertionError("should not happen", e);
+        }
+    }
+
+    @SuppressWarnings("removal")
+    static VarHandle privilegedFindVarHandle(MethodHandles.Lookup lookup, Class<?> cls, String name, Class<?> type)
+        throws PrivilegedActionException {
+        PrivilegedExceptionAction<VarHandle> pa = () -> lookup.findVarHandle(cls, name, type);
+        return AccessController.doPrivileged(pa);
+    }
+
+    private IndexInputUtils() {}
+
+    static MemorySegment segmentSlice(IndexInput input, long pos, int length) {
+        if (MS_MSINDEX_CLS.isAssignableFrom(input.getClass())) {
+            pos += offset(input);
+        }
+        final int si = (int) (pos >> chunkSizePower(input));
+        final MemorySegment seg = segmentArray(input)[si];
+        long offset = pos & chunkSizeMask(input);
+        if (checkIndex(offset + length, seg.byteSize() + 1)) {
+            return seg.asSlice(offset, length);
+        }
+        return null;
+    }
```

Comment:
This will return `null` if the position and length pass between segments? I admittedly know very little about how memory segments work, but since Lucene chunks (16GB i think?) this in various sizes, my concern would be a vector that straddles a memory segment chunk.

Reply:
> This will return `null` if the position and length pass between segments?

Correct.

> I admittedly know very little about how memory segments work, but since Lucene chunks (16GB i think?

Yes, 16 GB by default.

> ) this in various sizes, my concern would be a vector that straddles a memory segment chunk.

Yeah a vector can infrequently straddle, and this is accounted for and tested. In the straddling situation we just fallback to copying on-heap and doing the regular Lucene/Panama thing.  I was going to try to optimise this situation, but felt that it was not worth it, since it's so infrequent.

Reply:
> I was going to try to optimise this situation, but felt that it was not worth it, since it's so infrequent.

OK, glad we got it covered :)